### ISSUE TITLE
analysis-icu (icu_normalizer) throw java.lang.StackOverflowError.

### ISSUE THREAD

**Issue by mitoma** [elastic/elasticsearch#26474]
**Elasticsearch version** : 5.5.2
**Plugins installed**: [analysis-icu]
**JVM version** : 1.8.0_131
**OS version** : Ubuntu 16.04.3 LTS
**Description of the problem including expected versus actual behavior**:

Actual : icu_normalizer throw `java.lang.StackOverflowError`. and server down.
Expected : Server is should not crash.

**Steps to reproduce**:

this script is complete scenario.

```sh
#!/bin/bash
#
# setup
# $ tar xzvf elasticsearch-5.5.2.tar.gz
# $ cd elasticsearch-5.5.2/
# $ ./bin/elasticsearch-plugin install analysis-icu
# $ ./bin/elasticsearch &
#
# crash test
# $ ./es_crash_poc.sh
#

ELASTICSEARCH_HOST=localhost
ELASTICSEARCH_HTTP_PORT=9200

read -d '' CRASH_TYPE <<EOF
{
  "aliases": {},
  "mappings": {
    "records": {
      "properties": {
        "field1": {
          "type": "text",
          "analyzer": "crash_analyzer"
        }
      }
    }
  },
  "settings": {
    "index": {
      "analysis": {
        "analyzer": {
          "crash_analyzer": {
            "char_filter": [
              "icu_normalizer"
            ],
            "tokenizer": "icu_tokenizer"
          }
        }
      }
    }
  }
}
EOF

# create index
curl -X PUT http://${ELASTICSEARCH_HOST}:${ELASTICSEARCH_HTTP_PORT}/es_crash_index -d "$CRASH_TYPE"

read -d '' CRASH_DOC <<EOF
{"field1":"`seq 1000000 | sed -e 's/.*/a/' | tr -d "\n"`"}
EOF

echo "$CRASH_DOC" > crash_doc.txt

# this insert kill a server.
curl -X PUT http://${ELASTICSEARCH_HOST}:${ELASTICSEARCH_HTTP_PORT}/es_crash_index/records/1 -d "@crash_doc.txt"
```

**Provide logs**:

```txt
[2017-09-01T22:42:31,870][INFO ][o.e.c.m.MetaDataCreateIndexService] [31T0CIp] [es_crash_index] creating index, cause [api], templates [], shards [5]/[1], mappings [records]
[2017-09-01T22:42:33,153][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [] fatal error in thread [elasticsearch[31T0CIp][index][T#1]], exiting
java.lang.StackOverflowError: null
        at org.apache.lucene.analysis.ReusableStringReader.read(ReusableStringReader.java:47) ~[lucene-core-6.3.0.jar:6.3.0 a66a44513ee8191e25b477372094bfa846450316 - shalin - 2016-11-02 19:47:11]
        at java.io.Reader.read(Reader.java:140) ~[?:1.8.0_121]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:100) ~[?:?]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:113) ~[?:?]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:113) ~[?:?]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:113) ~[?:?]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:113) ~[?:?]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:113) ~[?:?]
        at org.apache.lucene.analysis.icu.ICUNormalizer2CharFilter.readInputToBuffer(ICUNormalizer2CharFilter.java:113) ~[?:?]
--snip--
```

**Comment by jpountz** [comment]
Thank you for the report. I could reproduce the issue and opened https://issues.apache.org/jira/browse/LUCENE-7956.

**Comment by mitoma** [comment]
Thank you! I will wait.

**Comment by jpountz** [comment]
The fix has been merged and will be in Lucene 7.0 and thus in Elasticsearch 6.0 as well. I's not clear whether we'll be able to have it into Elasticsearch 5.x however, maybe not.
